cdap-ui/app/cdap/components/ConfigurationGroup/utilities/index.ts

```diff
@@ -119,3 +135,153 @@ export function isMacro(value) {
 
   return beginChar && endingChar;
 }
+
+export function constructErrors(failures) {
+  // Structure is
+  // propertyErrors = {
+  //   prop1Name: [{
+  //     msg: 'error msg',
+  //   }],
+  //   prop2Name: [{
+  //     element: 'keyValuePairRow2',
+  //     msg:'error in second row of keyvalue pair field'
+  //   }]
+  // }
+  // prop1Name identifies widgets, element is used to identify sub-elements
+  // like rows in multi row widget etc.
+  const propertyErrors: IPropErrors = {};
+  // Structure for inputSchemaErrors and outputSchemaErrors :
+  // inputSchemaErrors = {
+  //   'default':{
+  //     field1:'error msg',
+  //     field2: 'error msg'
+  //   },
+  //  'stage1':{
+  //    field1:'error msg',
+  //    field2:'error msg'
+  //  }
+  // }
+  const inputSchemaErrors: IInputSchemaErrors = {};
+  const outputSchemaErrors: IOutputSchemaErrors = {};
+
+  failures.forEach((failure) => {
+    const errorMsg = constructErrorMessage(failure.message, failure.correctiveAction);
+    if (failure.causes) {
+      failure.causes.forEach((cause) => {
+        if (cause.attributes) {
+          const { stageConfig, inputField, outputField } = cause.attributes;
+          // stageConfig is the field used to identify that the error belongs
+          // to plugin properties.
+          if (stageConfig) {
+            const err: IErrorObj = { msg: errorMsg };
+            const { configElement } = cause.attributes;
+            if (configElement) {
+              // If configElement is present, it represents that the error is
+              // in a complex widget i.e second row in a multirow widget.
+              err.element = configElement;
+            }
+            // gathering errors in a widget
+            if (propertyErrors.hasOwnProperty(stageConfig)) {
+              propertyErrors[stageConfig].push(err);
+            } else {
+              propertyErrors[stageConfig] = [err];
+            }
+          }
+          // inputField is the field used to identify that the error occurs in
+          // inputSchema
+          else if (inputField) {
+            const { inputStage, stage } = cause.attributes;
+            const errorStage = inputStage ? inputStage : stage;
+            insertIntoObj(inputSchemaErrors, errorStage, inputField, errorMsg);
+          }
+          // outputField is the field used to identify that the error occurs
+          // in outputSchema
+          else if (outputField) {
+            const { outputPort } = cause.attributes;
+            // noSchemaSection is used when there are no ports in op schema
+            const errorStage = outputPort ? outputPort : 'noSchemaSection';
+            insertIntoObj(outputSchemaErrors, errorStage, outputField, errorMsg);
+          } else {
```

Comment:
nit: For readability can we have each of this condition as separate `if` conditions? Right now this `else` gets repeated outside as well.

Reply:
Will do it with the follow up work.